Migrate web site from Next to Vite

rwblair opened the pull request:
Help match the openneuro build environment more closely.

I didn't need to use any of the workarounds in https://github.com/OpenNeuroOrg/openneuro/blob/master/packages/openneuro-app/vite.config.js to allow vite build to go off. Mosttly stayed with defaults from vite.

nellh commented:
> I didn't need to use any of the workarounds in https://github.com/OpenNeuroOrg/openneuro/blob/master/packages/openneuro-app/vite.config.js to allow vite build to go off. Mosttly stayed with defaults from vite.

I wonder if that's fixes in Vite or because they are dependencies of dependencies when used in OpenNeuro?

rwblair commented:
vite 2.8 vs vite 2.3.1 in openneuro, could be. The tsconfigs between ON and validator are slightly different so not sure if that affects it.

rwblair commented:
I'm replicating those locally outside of docker, not sure whats up.

rwblair commented:
Right now the two packages throwing esbuild errors for me are nifti-js and jshint. nifti-js is from an unused require('assert') and jshint has the same import, but assert is being used.

nifti-js hasn't been updated since 2016 :(

rwblair commented:
To get the esbuild to work I cloned [nifti-js](https://github.com/rwblair/nifti-js) and removed the unused blocking import. Keeping it at my rwblair account is no good, should I push a copy to bids-standard, or should we be trying something like monkey patching? Project is effectively abandoned.

I also removed jshint from the config json parser. I have it return the json.parse error message instead, so this is considerably less verbose, not sure how much we care about this feature.

effigies commented:
Maybe try opening a PR against nifti-js before we fork?

Could we bundle it into our source tree (maybe as a submodule?) until it's merged and released upstream or we decide to officially fork?

nellh commented:
[nifti-reader-js](https://github.com/rii-mango/NIFTI-Reader-JS) may be a good alternative, it's used by NiiVue so at least on the OpenNeuro side it could slightly reduce dependency weight.

rwblair commented:
nifti-header-js uses different names for header fields than the spec and nifti-js so that caused a bit of silliness (pixDims vs pixdim and dims vs dim). I should of just normalized the field names to match the spec in the parseHeader function, instead I changed them to match the libraries usage. Fun fact nifti-js was overwriting the dim array length property to match dim[0].

rwblair commented:
Unfortunately something like this won't work since it doesn't kill the old dev instance:
https://www.npmjs.com/package/npm-watch

Vites watch can't obviously kick off the validator rebuild as far as I can tell. Not sure about npm script interactions in that config:
https://vitejs.dev/config/#server-watch

nellh commented on package.json:
```diff
@@ -11,7 +11,7 @@
     "postversion": "lerna version prerelease --preid dev --ignore-scripts --yes --force-publish=*",
     "prepublishOnly": "npm run prepublishOnly -w bids-validator",
-    "web-dev": "cd bids-validator-web && npm run dev",
+    "web-dev": "npm run build -w bids-validator && npm run dev -w bids-validator-web",
```
I don't think this is the right fix, since it dev mode still doesn't pick up new changes.

This change to bids-validator/package.json gets dev mode running but runs into some other issues.
```json
  "exports": {
    ".": {
      "import": "./index.js",
      "require": "./dist/commonjs/index.js"
    },
    "./cli": {
      "require": "./dist/commonjs/cli.js"
    },
    "./utils/consoleFormat": {
      "require": "./dist/commonjs/consoleFormat.js"
    }
  },
```

rwblair commented:
@nellh I couldn't figure out how to get it to go without using the openneuro shims, only needed two of them.

As best I can figure in the bids-validator esbuild was taking queues from the browser field in the package.json field, and there is no way to pass that information outside of packge.json to esbuild.

rwblair commented:
rwblair/bids-validator#3
An alternative idea that goes back to using dist/esm/index.js, and enabling esbuild's built in watch functionality on run web-dev, and then having the esbuild and vite dev run simultaneously using a package called concurrently.
